seismic_hazard_analysis/nshm_2022/oq_interface.py

```diff
@@ -0,0 +1,236 @@
+
+    # Split the realizations into batches
+    # for loading the hazard curve realisations
+    # as loading all of them at once is extremely slow
+    n_batches = int(np.ceil(n_rels / 16))
+    batches = np.array_split(weights_df.index.values, n_batches)
+
+    start = time.time()
+    if n_procs == 1:
+        hcurves = []
+        for batch in tqdm(batches):
+            hcurves.append(_get_hazard_curves(calc_id, batch, n_ims, n_im_levels))
+    else:
+        with mp.Pool(n_procs) as pool:
+            hcurves = pool.starmap(
+                _get_hazard_curves,
+                [(calc_id, batch, n_ims, n_im_levels) for batch in batches],
+            )
+    print(f"Took: {time.time() - start} to get {weights_df.shape[0]} hazard curves")
+    hcurves = np.concatenate(hcurves, axis=0)
+
+    # Compute the statistics
+
+    weights = weights_df.values
+    for i, cur_im in tqdm(enumerate(ims), total=n_ims, desc="Processing IMs"):
+        cur_data = hcurves[:, i, :]
+
+        # Calculate the mean hazard
+        cur_mean = np.average(cur_data, weights=weights, axis=0)
+
+        # Calculate the quantiles
+        sort_ind = np.argsort(cur_data, axis=0)
+        cur_data = np.take_along_axis(cur_data, sort_ind, axis=0)
+        cur_weights = np.tile(weights[:, None], (1, n_im_levels))
+        cur_weights = np.take_along_axis(cur_weights, sort_ind, axis=0)
+
+        cur_quantile_results = utils.query_non_parametric_multi_cdf_invs(
+            quantiles, cur_data.T, np.cumsum(cur_weights.T, axis=1)
+        )
+
+        # Combine and write
+        cur_df = pd.DataFrame(
+            data=np.stack([cur_mean] + list(cur_quantile_results), axis=1),
+            columns=["mean"] + [f"quantile_{cur_qt}" for cur_qt in quantiles],
+            index=im_levels[cur_im],
+        )
+
+        cur_df.to_csv(
+            output_dir
+            / f"{utils.get_im_file_format(get_im_name(cur_im))}_statistics.csv",
+            index_label="im_level",
+        )
+
+
+def _get_hazard_curves(
+    calc_id: int,
+    rel_ordinals: np.ndarray[int],
+    n_ims: int,
+    n_im_levels: int,
+) -> np.ndarray:
+    """Helper function to get hazard curves for a batch of realizations."""
+    with Extractor(calc_id) as ex:
+        hcurves = np.full(
+            (rel_ordinals.size, n_ims, n_im_levels), np.nan, dtype=np.float32
+        )
+        for i, cur_rel_ordinal in enumerate(rel_ordinals):
+            hcurves[i, :, :] = ex.get(f"hcurves?kind=rlz-{cur_rel_ordinal}&site_id=0")[
+                f"rlz-{cur_rel_ordinal:03}"
+            ][0, :, :]
+
+    return hcurves
+
+
+def get_im_name(oq_im: str) -> str:
+    """
+    Get the name of the IM from the OpenQuake IM name.
+
+    Parameters
+    ----------
+    im: str
+        The OpenQuake IM name.
+
+    Returns
+    -------
+    str
+        The name of the IM.
+    """
+    if oq_im.startswith("SA"):
+        return oq_im.replace("SA", "pSA").replace("(", "_").replace(")", "")
+
+    return oq_im
```

Comment:
Could be simplified with some regex:

```suggestion
    return re.sub(r"^SA\((.*)\)", r"pSA_\1", oq_im)
```